### PR TITLE
Plugin data object

### PR DESCRIPTION
**Pull request by miladh**
TODO:
- [x] Use `AttributeData` object in `attribute.py`
- [x] Update the existing plugins to use `AttributeData`
- [x] Use `AttributeData` and `DatasetData` in `prepare_read()`
- [x] Write meta data from attributes in exdir.yaml files

### CODE REVIEW

File: exdir/plugin_interface/plugin_interface.py [hunk 1]
```diff
@@ -31,8 +31,21 @@ def __init__(self, name, dataset_plugins=None, attribute_plugins=None,
                 setattr(plugin, "_plugin_module", self)
 
 
+class DatasetData:
+    def __init__(self, data, attrs, meta):
```
Comment (simetenn) [inline exdir/plugin_interface/plugin_interface.py]: We should probably state what `data`, `attrs` and `meta` are.

File: exdir/plugin_interface/plugin_interface.py [hunk 2]
```diff
@@ -31,8 +31,21 @@ def __init__(self, name, dataset_plugins=None, attribute_plugins=None,
...
+        self.data = data
+        self.attrs = attrs
+        self.meta = meta
+
+
+class AttributeData:
+    def __init__(self, attrs, meta):
```
Comment (simetenn) [inline exdir/plugin_interface/plugin_interface.py]: We should probably state the what `attrs` and `meta` are.

**Review by simetenn — APPROVED** [review]
This seems good to me.

File: exdir/plugin_interface/plugin_interface.py (hunk 1 above)
Comment (miladh) [reply exdir/plugin_interface/plugin_interface.py]: fixed in 4b6389028509145f40be518034444ee3dd15f710

File: exdir/plugin_interface/plugin_interface.py (hunk 2 above)
Comment (miladh) [reply exdir/plugin_interface/plugin_interface.py]: fixed in 4b6389028509145f40be518034444ee3dd15f710

File: exdir/core/attribute.py [hunk 3]
```diff
@@ -22,95 +22,105 @@ def __init__(self, parent, mode, io_mode, path=None, plugin_manager=None):
...
+                                                              meta=meta)
+
         for plugin in self.plugin_manager.attribute_plugins.write_order:
-            meta_data = plugin.prepare_write(meta_data)
+            attribute_data = plugin.prepare_write(attribute_data)
+
+            if "required" in attribute_data.meta and attribute_data.meta["required"] is True:
+                meta[plugin._plugin_module.name] = attribute_data.meta
+
+        # QUESTION: Should we add plugin meta data to parent.meta here?
```
Comment (miladh) [inline exdir/core/attribute.py]: We need to decide how to include meta information (e.g. `required: True`) about an attribute plugin. One option is to update `parent.meta["plugins"]`. What do you think @dragly?

Comment (dragly) [reply exdir/core/attribute.py]: Yes, I think so. I'll look into it now.

**Pull request merged**